docs/database.rst

```diff
@@ -0,0 +1,231 @@
+This requires using the `SQLAlchemy 2.0 API <https://docs.sqlalchemy.org/en/14/tutorial/index.html>`__, which is somewhat different than the older API.
+Safir uses the `asyncpg`_ PostgreSQL database driver.
+
+Database support in Safir is optional.
+To use it, depend on ``safir[db]``.
+
+Initializing a database
+=======================
+
+Safir supports simple initialization of an empty database with a schema provided by the application.
+The application may also request a database reset, which will drop and recreate all of the tables in the schema.
+More complex database schema upgrades are not supported by Safir.
+If those are required, consider using `Alembic <https://alembic.sqlalchemy.org/en/latest/>`__.
+
+Database initialization in Safir assumes that the application has defined the database schema via the SQLAlchemy ORM.
+The recommended way to do this is to add a ``schema`` directory to the application containing the table definitions.
+In the file ``schema/base.py``, define the SQLAlchemy declarative base:
+
+.. code-block:: python
+
+   from sqlalchemy.orm import declarative_base
+
+   Base = declarative_base()
+
+In other files in that directory, define the database tables using the normal SQLAlchemy ORM syntax, one table per file.
+Each database table definition must inherit from ``Base``, imported from ``.base``.
+In ``schema/__init__.py``, import the table definitions from all of the files in the directory, as well as the ``Base`` variable, and export them using ``__all__``.
+
+Then, Safir can be used to initialize the database with code like:
+
+.. code-block:: python
+
+   import structlog
+   from safir.database import initialize_database
+
+   from .config import config
+   from .schema import Base
+
+
+   logger = structlog.get_logger(config.logger_name)
+   engine = await initialize_database(
+       config.database_url,
+       config.database_password,
+       logger,
+       schema=Base.metadata,
+   )
+   await engine.dispose()
+
+(Although not shown here, this must be done inside an ``async`` function.)
+This code assumes that ``.config`` provides a ``config`` object that contains the settings for the application, including the database URL and password as well as the normal Safir configuration settings.
+
+If it receives a connection error from the database, Safir will attempt the initialization five times, two seconds apart, to allow time for networking or a database proxy to start.
+
+To drop and recreate all of the tables, pass the ``reset=True`` option to `~safir.database.initialize_database`.
+
+Note that `~safir.database.initialize_database` returns a `~sqlalchemy.ext.asyncio.AsyncEngine` object for the newly-initialized database.
+This can be used to perform any further application-specific database initialization that is required, such as adding default table entries.
+Put any such code before the ``await engine.dispose()`` call.
+
+.. _async-db-session:
+
+Creating an async database session
```

Comment:
So I think the main thing we want to get out of this section is how to write a `check_database` function that's run during application start-up, and then parenthetically say that `create_async_session` can also be used to create a database session outside a request.

So I'd be tempted to reformat this section around that `check_database()` task, and then if we need an explicit section about getting an async session outside a request dependency, maybe put that section after "Using a database session in request handlers", next to "Creating a sync database session".

I think this keeps the most important and actionable docs first.

Reply:
I actually don't think you should write a `check_database` call at all, and instead the application startup script should call `app init` before starting the app with uvicorn (which will, as a side effect, test database connectivity), but this should definitely be documented. I'm going to tackle that in a later diff when the Click stuff is introduced, so for this I think I'm going to move the database dependency up as the normal way to interact with the database and rework this section to be clear that this is for non-FastAPI uses of the database, such as cron jobs, background processing, or other supporting code.